add a postprocessor for the volumetric strain rate

jdannberg opened the pull request:
This adds a visualization postprocessor for the volumetric strain rate, which is in particular useful in compressible models or models with melt transport.

tjhei commented on source/postprocess/visualization/volumetric_strain_rate.cc:
```diff
@@ -0,0 +1,88 @@
+*/
+
+
+#include "aspect/postprocess/visualization/volumetric_strain_rate.h"
```
make this a system include (using ``<...>``)

tjhei commented on source/postprocess/visualization/volumetric_strain_rate.cc:
```diff
@@ -0,0 +1,88 @@
+                                                  "for the volumetric strain rate, i.e., for the quantity "
+                                                  "$\\nabla\\mathbf u$. "
+                                                  "This should be zero (up to the solver tolerance) in incompressible "
+                                                  "convection models, but can be nonnzero in compressible models and "
```
typo: nonnzero

jdannberg commented:
Fixed.

bangerth commented on source/postprocess/visualization/volumetric_strain_rate.cc:
```diff
@@ -0,0 +1,88 @@
+                                                  "volumetric strain rate",
+                                                  "A visualization output object that generates output "
+                                                  "for the volumetric strain rate, i.e., for the quantity "
+                                                  "$\\nabla\\mathbf u$. "
```
are you missing a cdot? maybe write this as `\\nabla\\cdot\\mathbf u = \\textrm{div}\; \\mathbf u = \textrm{trace}\; \\varepsilon(\\mathbf u)`

bangerth commented on include/aspect/postprocess/visualization/volumetric_strain_rate.h:
```diff
@@ -0,0 +1,62 @@
+       * A class derived from DataPostprocessor that takes an output vector
+       * and computes a variable that represents the volumetric strain rate at every
+       * point. The volumetric strain rate is defined as $\nabla \mathbf u$
+       * (or the trace of the strain rate tensor).
```
same comment as I leave below

bangerth commented on source/postprocess/visualization/volumetric_strain_rate.cc:
```diff
@@ -0,0 +1,88 @@
+                                                  "$\\nabla\\mathbf u$. "
+                                                  "This should be zero (up to the solver tolerance) in incompressible "
+                                                  "convection models, but can be non-zero in compressible models and "
+                                                  "models with melt transport.")
```
that's not actually true. it is zero for the *exact* solution, but for the numerical solution it is not in general. All we can say is that the divergence is zero in some average sense (namely when multiplied by a test function and integrated over) but it is not zero at every point.

bangerth commented on tests/compressibility_iterated_stokes.prm:
```diff
@@ -108,11 +108,11 @@ subsection Boundary velocity model
   subsection Visualization
 
-    set List of output variables      = density
+    set List of output variables      = density, volumetric strain rate
```
just copy/clone the test. we like having tests that test one particular aspect, rather than many things at the same time

jdannberg commented:
I think I addressed all of your comments. For the work where we compare the different formulations, we need rho * div u, which is a slightly different postprocessor (as that requires evaluating the material model), which probably also does not have that many uses, so we only implemented that as a plugin (not as one of the regular postprocessors).

bangerth commented on source/postprocess/visualization/volumetric_strain_rate.cc:
```diff
@@ -0,0 +1,89 @@
+            for (unsigned int d=0; d<dim; ++d)
+              {
+                grad_u[d] = input_data.solution_gradients[q][d];
+                div_u += grad_u[d][d];
```
I missed this before -- you're assigning `grad_u[d]` once, and then you use it once. It is, in principle, not necessary to even put this into a variable. I think I'd just get rid of the `grad_u` variable altogether.

jdannberg replied:
I can do that, I just thought if I write it like this, the code is easier to understand.

bangerth replied:
Sure, but in that case I'd first fill `grad_u`, and then compute the result by saying `trace(grad_u)`.

jdannberg replied:
I like that, I didn't think of writing it that way. I changed it.
